packages/authentication/src/strategy-adapter.ts

```diff
@@ -0,0 +1,81 @@
+// Copyright IBM Corp. 2013,2017. All Rights Reserved.
+// Node module: loopback
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+import * as http from 'http';
+import {ServerRequest as Request} from 'http';
+import {HttpErrorNotFound} from '@loopback/core';
+import {ParsedRequest} from '@loopback/core';
+
+/**
+ * Interface definition of a passport strategy.
+ */
+export interface Strategy {
+  authenticate(req: http.ServerRequest): void;
+}
+
+/**
+ * Shimmed Request to satisfy express requirements of passport strategies.
+ */
+export class ShimRequest {
+  headers: Object;
+  query: Object;
+  url: string;
+  path: string;
+  method: string;
+  constructor(request?: ParsedRequest) {
+    if (request) {
+      this.headers = request.headers;
+      this.query = request.query;
+      this.url = request.url;
+      this.path = request.path;
+      this.method = request.method;
+    }
+  }
+}
+
+/**
+ * Adapter class to invoke a passport strategy.
+ * Instance is created by passing a passport strategy in the constructor
+ */
+export class StrategyAdapter {
+  private strategyCtor: Strategy;
+
+  constructor(strategy: Strategy) {
+    this.strategyCtor = strategy;
+  }
+
+  /**
+   * The function to invoke the contained passport strategy.
+   *     1. Create an instance of the strategy
+   *     2. add success and failure state handlers
+   *     3. authenticate using the strategy
+   * @param req {http.ServerRequest} The incoming request.
+   */
+  authenticate(req: ParsedRequest) {
+    const shimReq = new ShimRequest(req);
+    return new Promise<Object>((resolve, reject) => {
+      // create an instance of the strategy
+      const strategy = Object.create(this.strategyCtor);
+      const self = this;
+
+      // add success state handler to strategy instance
+      strategy.success = function(user: object) {
+        resolve(user);
+      };
+
+      // add failure state handler to strategy instance
+      strategy.fail = function(challenge: string) {
+        reject(new HttpErrorNotFound(challenge));
```

Comment:
This is returning 404 Not Found, are we sure that's the correct status code? My expectation is that the server should send 401 Not Authorized with WWW-Authenticate header field containing a challenge applicable to the requested resource.

@raymondfeng do you have any opinion on this?